midend/replaceSelectRange.cpp

```diff
@@ -21,7 +21,7 @@
 namespace P4 {
 
 std::vector<const IR::Mask *>
-DoReplaceSelectRange::rangeToMasks(const IR::Range *r) {
+DoReplaceSelectRange::rangeToMasks(const IR::Range *r, int keyIndex) {
```

Comment:
I think it's actually nicer to return a list of vectors for the result.
one in the common case, two if the range contains 0.
the keyIndex and signedIndicesToReplace make the data flow much harder to understand.

Reply:
I agree that `signedIndicesToReplace` does make the flow harder to understand. But this is mainly because it is indeed complex -- the indices are detected when range is expanded, but they are used in the corresponding select through which we backtrack later (making use of `postorder`). This needs to be done this way as at the time we encounter the select first, we don't know which values will contain ranges and when working with ranges we cannot replace a parent node (at least I don't think so, I could be wrong since I am a beginner to p4c). The only alternative to indices I see would be to remember the actual corresponding select expressions, but that in my opinion improves nothing and could lead to danger in case of some weird select with duplicated keys:

```
select (foo.a, foo.b, foo.a) {
   -5..5, 42, _ : ...
   _, 16, -32000 : ...
}
```

In this (rather pathological) case, we need to insert the bitcast only to the first occurence of `foo.a` not to both instances of `foo.a` expression in select.

---

As for the vector-of-vectors. I don't think this is a good idea as returning a vector of vectors implies that there can be any number of vectors, including 0 and more than 2. Therefore I think that would be bad design of the function as the type would allow too many invalid possibilities. It could make sense to return pair (isSigned, vectors) and resolve the index one level up in the postorder of mask.

---

For now, I have just added a comment about the indices into the class.